subprojects/core/src/main/java/org/gradle/internal/build/BuildOperationFiringBuildWorkPreparer.java

```diff
@@ -120,104 +139,87 @@ void populateTaskGraph() {
             delegate.populateWorkGraph(gradle, plan, action);
         }
 
+        @Nonnull
         @Override
         public BuildOperationDescriptor.Builder description() {
             //noinspection Convert2Lambda
             return BuildOperationDescriptor.displayName(gradle.contextualize("Calculate task graph"))
-                .details(new CalculateTaskGraphBuildOperationType.Details() {
+                .details(new Details() {
                     @Override
                     public String getBuildPath() {
                         return gradle.getIdentityPath().getPath();
                     }
                 });
         }
-    }
 
-    private static List<CalculateTaskGraphBuildOperationType.PlannedTask> toPlannedTasks(QueryableExecutionPlan.ScheduledNodes scheduledWork) {
-        List<CalculateTaskGraphBuildOperationType.PlannedTask> tasks = new ArrayList<>();
-        scheduledWork.visitNodes(nodes -> {
-            for (Node node : nodes) {
-                if (node instanceof LocalTaskNode) {
-                    tasks.add(toPlannedTask((LocalTaskNode) node));
+        private List<PlannedNode> toPlannedNodes(QueryableExecutionPlan.ScheduledNodes scheduledWork) {
+            List<PlannedNode> plannedNodes = new ArrayList<>();
+            scheduledWork.visitNodes(nodes -> {
+                for (Node node : nodes) {
+                    ToPlannedNodeConverter converter = getConverter(node);
+                    if (converter != null && converter.isInSamePlan(node)) {
+                        PlannedNode plannedNode = converter.convert(node, this::findDependencies);
```

Comment:
🤔 We should probably think about this API again. I don't like the non-atomicity we have now. Though I also don't like it returning `null`.